core/embed/rust/src/ui/display/tjpgd.rs

```diff
@@ -0,0 +1,1561 @@
+const HUFF_LEN: u32 = 1 << HUFF_BIT;
+const HUFF_MASK: u32 = HUFF_LEN - 1;
+
+const NUM_DEQUANTIZER_TABLES: usize = 4;
+
+#[derive(PartialEq, Eq)]
+pub enum Error {
+    /// Interrupted by output function, call `JDEC::decomp` to continue.
+    Interrupted,
+    /// Device error or wrong termination of input stream.
+    Input,
+    /// Insufficient memory pool for the image.
+    MemoryPool,
+    /// Insufficient stream input buffer.
+    MemoryInput,
+    /// Parameter error.
+    Parameter,
+    /// Data format error (may be broken data).
+    InvalidData,
+    /// Not supported JPEG standard.
+    UnsupportedJpeg,
+}
+
+pub struct JDEC<'i, 'p> {
+    dctr: usize,
+    dptr: usize,
+    inbuf: &'p mut [u8],
+    dbit: u8,
+    scale: u8,
+    msx: u8,
+    msy: u8,
+    qtid: [u8; 3],
+    ncomp: u8,
+    dcv: [i16; 3],
+    nrst: u16,
+    rst: u16,
+    rsc: u16,
+    width: u16,
+    height: u16,
+    huffbits: [[&'p mut [u8]; 2]; 2],
+    huffcode: [[&'p mut [u16]; 2]; 2],
+    huffcode_len: [[usize; 2]; 2],
+    huffdata: [[&'p mut [u8]; 2]; 2],
+    qttbl: [&'p mut [i32]; 4],
+    wreg: u32,
+    marker: u8,
+    longofs: [[u8; 2]; 2],
+    hufflut_ac: [&'p mut [u16]; 2],
+    hufflut_dc: [&'p mut [u8]; 2],
+    workbuf: &'p mut [i32],
+    mcubuf: &'p mut [i16],
+    pool: &'p mut [u8],
+    input_func: &'i mut dyn JpegInput,
+}
+
+/// Zigzag-order to raster-order conversion table
+#[rustfmt::skip]
+const ZIG: [u8; 64] = [
+     0,  1,  8, 16,  9,  2,  3, 10, 17, 24, 32, 25, 18, 11,  4,  5,
+    12, 19, 26, 33, 40, 48, 41, 34, 27, 20, 13,  6,  7, 14, 21, 28,
+    35, 42, 49, 56, 57, 50, 43, 36, 29, 22, 15, 23, 30, 37, 44, 51,
+    58, 59, 52, 45, 38, 31, 39, 46, 53, 60, 61, 54, 47, 55, 62, 63,
+];
+
+macro_rules! f {
+    ($num:expr) => {{
+        ($num * 8192_f64) as u16
+    }};
+}
+
+/// Input scale factor of Arai algorithm
+/// (scaled up 16 bits for fixed point operations)
+#[rustfmt::skip]
+const IPSF: [u16; 64] = [
+    f!(1.00000), f!(1.38704), f!(1.30656),       f!(1.17588), f!(1.00000), f!(0.78570), f!(0.54120),       f!(0.27590),
+    f!(1.38704), f!(1.92388), f!(1.81226),       f!(1.63099), f!(1.38704), f!(1.08979), f!(0.75066),       f!(0.38268),
+    f!(1.30656), f!(1.81226), f!(1.70711),       f!(1.53636), f!(1.30656), f!(1.02656), f!(FRAC_1_SQRT_2), f!(0.36048),
+    f!(1.17588), f!(1.63099), f!(1.53636),       f!(1.38268), f!(1.17588), f!(0.92388), f!(0.63638),       f!(0.32442),
+    f!(1.00000), f!(1.38704), f!(1.30656),       f!(1.17588), f!(1.00000), f!(0.78570), f!(0.54120),       f!(0.27590),
+    f!(0.78570), f!(1.08979), f!(1.02656),       f!(0.92388), f!(0.78570), f!(0.61732), f!(0.42522),       f!(0.21677),
+    f!(0.54120), f!(0.75066), f!(FRAC_1_SQRT_2), f!(0.63638), f!(0.54120), f!(0.42522), f!(0.29290),       f!(0.14932),
+    f!(0.27590), f!(0.38268), f!(0.36048),       f!(0.32442), f!(0.27590), f!(0.21678), f!(0.14932),       f!(0.07612),
+];
+
+fn reinterpret_slice<T>(slice: &[i32]) -> &[T] {
```

Comment:
could we replace those calls with `slice.align_to()` / `align_to_mut`? that way it would be much clearer, at usage site, that the usage is safe
(fwiw as written these functions are very unsafe because they _internally_ rely on the type being integer

Reply:
https://github.com/trezor/trezor-firmware/pull/2673/commits/dc8b6172198532b1920a3089e3cfb674855c71a8